c2rust-analyze/src/main.rs

```diff
@@ -150,6 +150,18 @@ fn run(tcx: TyCtxt) {
         gacx.fn_sigs.insert(ldid.to_def_id(), lsig);
     }
 
+    // Label the field types of each struct.
+    for ldid in tcx.hir_crate_items(()).definitions() {
+        let did = ldid.to_def_id();
+        if tcx.def_kind(did) != DefKind::Struct {
+            continue;
+        }
+        for field in tcx.adt_def(did).all_fields() {
```

Comment:
Where's the definition of `tcx.adt_def`?  I can't find it buried within the query system.  Why does it not return an `Option`?

Reply:
Looks like the real definition is [here](https://doc.rust-lang.org/stable/nightly-rustc/rustc_typeck/collect/fn.adt_def.html).  I found this by searching the docs for things named `adt_def` and scrolling down until I found a free function by that name (usually query implementations are defined as free functions with the same name as the query).  It happens to be pretty far down the list in this case.

>Why does it not return an Option?

It panics on failure instead.  You're supposed to ensure the precondition some other way before calling (as we do by checking `tcx.def_kind(did)`).